### ISSUE TITLE
Animation name not applied in Piskel

### ISSUE THREAD

**Issue by PascalLadalle** [4ian/GDevelop#1751]
## Describe the bug
It is known that when you duplicate an object and then edit its sprite in Piskel, you're unable to rename it inside Piskel, so it overwrites the sprite of the original object.
What I noticed now is that if you rename the animation before opening it in Piskel, the new name appears in the file address bar, but when you save, the original file is still overwritten.

## To Reproduce
1. Duplicate a sprite.
2. Rename its first animation.
3. In Piskel, edit the sprite, see the new animation name written on top, and save.
4. Hover your mouse on a sprite of the animation to see that the old filename was kept, or close the window and see that the original file has been updated like the duplicate.

## Other details
Beta 93 installed on Windows 10 64-bit

The workaround is to delete the animation and create a new one and copy/paste the old one inside (or open separately in Piskel, perhaps)

**Comment by PawBud** [comment]
Does the issue persist in GDevelop b94, The changelog doesn't say much about this issue hence I think it should persist but nonetheless I will be glad if you can just verify it once!

**Comment by PawBud** [comment]
![Screenshot from 2020-05-17 20-42-28](https://user-images.githubusercontent.com/48093317/82152461-f57afa00-987e-11ea-92db-c42a7ede0555.png)
![Screenshot from 2020-05-17 20-42-18](https://user-images.githubusercontent.com/48093317/82152463-f7dd5400-987e-11ea-9bd7-78d2cc184a19.png)

**Comment by PawBud** [comment]
update yes it does; I tried it on b89 and the  issue persists there too, I'll try to find a fix for this issue

**Comment by blurymind** [comment]
When you edit a sprite in Piskel and the sprite comes from gdevelop(exported via gdevelop, not piskel), piskel keeps the paths to each frame, so as to allow you to overwrite the same png files it used before (frame.originalPath). When you duplicate a sprite in gdevelop, piskel has no idea that has occurred, and since the duplicate sprite is using the same original paths to frames (png files) - the files will get overwritten. One way to avoid this is to change the export path to a subfolder

Remember that you may have created some new frames in piskel and end up having ones with original paths and ones with automatically created ones

**Comment by blurymind** [comment]
@PawBud I left you some comments on the closed pr you tagged me on to help you find out what happens to the animation name in piskel :)

**Comment by PawBud** [comment]
Thanks a lot, @blurymind; I am having a look at each one of them

**Comment by PawBud** [comment]
> One way to avoid this is to change the export path to a subfolder

This is the part that I am unable to figure out how to do.

**Comment by PawBud** [comment]
> If this is just to update the animation name, have a look at what is passed to the pathEditor when the animation sequence is opened in piskel. Find out why the name is not being applied

The animation name is being applied as said by Pascal:- 
> What I noticed now is that if you rename the animation before opening it in Piskel, the new name appears in the file address bar, but when you save, the original file is still overwritten.

uff, now I am confused about this, am I getting something wrong here @blurymind?

**Comment by PawBud** [comment]
> I suggest you study how piskel creates/updates these paths on the file metadata storing the piskel file. You can alter this file metadata both in piskel's bridge code and in gdevelop ide's code. If the user does something in the gd ide and piskel needs to know about it, you can update the metadata on the ide side.

Ok so all I need to do is update the file metadata to "assign" a new Path to the frame so that the original frame(in the original sprites) doesn't get overwritten, am I right?

**Comment by blurymind** [comment]
> > I suggest you study how piskel creates/updates these paths on the file metadata storing the piskel file. You can alter this file metadata both in piskel's bridge code and in gdevelop ide's code. If the user does something in the gd ide and piskel needs to know about it, you can update the metadata on the ide side.
> 
> Ok so all I need to do is update the file metadata to "assign" a new Path to the frame so that the original frame(in the original sprites) doesn't get overwritten, am I right?

Lets focus on the bug at hand.
Its about the update of the animation name - nothing to do with paths. It would be a higher risk PR to touch the code that touches on them

**Comment by PawBud** [comment]
> We need to first decide what we are fixing :)

This is exactly where even I am confused. I see your point indeed

**Comment by blurymind** [comment]
@PascalLadalle can you make a gif or a video of your steps. It could make it a bit more obvious what is wrong here :) 
try https://www.screentogif.com/ if you are on windows

**Comment by PascalLadalle** [comment]
1- When I duplicate the object:
![image](https://user-images.githubusercontent.com/25978830/82705972-f1811a80-9c78-11ea-9cc8-a78e5c50e619.png)
2- I change the name:
![image](https://user-images.githubusercontent.com/25978830/82705992-fc3baf80-9c78-11ea-929b-9411b29ec60e.png)
3- Inside Piskel:
![image](https://user-images.githubusercontent.com/25978830/82706023-0f4e7f80-9c79-11ea-8aec-556ce68e9326.png)
4- After saving:
![image](https://user-images.githubusercontent.com/25978830/82706138-491f8600-9c79-11ea-92ed-ca2db804b2cf.png)
4- The original ball has indeed been overwritten (painted red):
![image](https://user-images.githubusercontent.com/25978830/82706209-73714380-9c79-11ea-9cca-b0d534fd0bfc.png)

When I edit in Piskel, it grabs the new animation name I gave and writes it as target file (3), but when I save, it still saves to the original file name.
Hope it's clearer.

**Comment by blurymind** [comment]
The animation name does not affect the path if it's imported from gdevelops ide. Duplicating the game object will not duplicate the frame files, so overwriting them is expected behavior. If you would like to get the ball to write to different paths with your new animation name, reimport the frames, but do it using piskel

**Comment by PawBud** [comment]
> When you edit a sprite in Piskel and the sprite comes from gdevelop(imported via gdevelop, not piskel), piskel keeps the paths to each frame, so as to allow you to overwrite the same png files it used before (frame.originalPath). When you duplicate a sprite in gdevelop, piskel has no idea that has occurred, and since the duplicate sprite is using the same original paths to frames (png files) - the files will get overwritten. One way to avoid this is to change the export path to a subfolder
> 
> Remember that you may have created some new frames in piskel and end up having ones with original paths and ones with automatically created ones.
> 
> It's tricky, but basically we want to be able to preserve the ability to save over the same frame files when we edit them in piskel, but we also want to avoid letting the user going into a corner case where they do something to cause unexpected behaviour
> 
> @PawBud I suggest you study how piskel creates/updates these paths on the file metadata storing the piskel file. You can alter this file metadata both in piskel's bridge code and in gdevelop ide's code. If the user does something in the gd ide and piskel needs to know about it, you can update the metadata on the ide side.
> 
> So gdevelop - the game engine takes frame sequences, which are flat png files. In order to preserve layers, piskel saves its layered json string to the resource's metadata, then it creates/updates frame png files for the engine to consume and tells the IDE about them.
> 
> In any case, I havent looked at that code for quite a while, so please study it carefully before doing fixes.
> Be careful not to introduce cases where the user keeps creating new files on each edit, a case where files are made in the wrong path or files are overwriting other files unexpectedly.
> 
> If this is just to update the animation name, have a look at what is passed to the pathEditor when the animation sequence is opened in piskel. Find out why the name is not being applied

Uhmm @blurymind, does that mean that the aforementioned suggestions is not the solution for this problem?

**Comment by blurymind** [comment]
The only good solution imo is to somehow give to the user an option to make duplicates of the frames when creating a duplicate of an object.  When you click edit in piskel on a copy of an object, piskel has no way of knowing the object is a copy and gdevelop sends it the original frame paths the object is using. You can probably get around that by duplicating the frames  inside piskel. Not sure if we need to do anything on this bug actually. If his bug is files being overwritten, that's expected behavior not a bug

**Comment by PawBud** [comment]
Hmm, you are right, If this is a behavior then There is no 'problem/bug' to solve.

**Comment by PascalLadalle** [comment]
> Piskel has no way of knowing if you are working on the original or the duplicate,

I'm not saying it should.
The bug is that the **file name** shown in Piskel, this:
![image](https://user-images.githubusercontent.com/25978830/82739741-771ccd00-9d42-11ea-9c52-f32df9617cc5.png)
is **wrong** and misleading because it is not the **actual** filename.

You are being confused by my duplicate example, but the bug is not related to duplicating, it's just the most common use case. Just rename the animation of an object, click on Edit with Piskel, and the file path shown in Piskel is your new animation name instead of the actual file name.

**Comment by blurymind** [comment]
This is not the filename. This is the animation name. It takes it from gdevelops animation you clicked to edit.

Are you saying that piskel is not updating it when you change it in gdevelop? 

When you report a bug, focus on your bug. Don't throw in features you like to have, because it will just confuse developers

**Comment by PascalLadalle** [comment]
Yes, thanks to my testing explained above, I see that it's the animation name, but Piskel is showing a **full path**, so it looks like the actual file name which, again, is **misleading** for users.
We should either show the **real** path, either not show any path.

**Comment by blurymind** [comment]
Piskel uses that to show a name when you edit a second single file. It is made clear to the user by adding a .png to it.
It shows the animation name as part of the path when you edit or create an animation with it. When you do so, piskel does indeed use the name to create the new animation frames, so it is valid to show it there.

You are the single only person who keeps raising this as an issue and repeating that it's confusing. I am sorry but I don't think it's an issue and you have not so far offered any solutions, just confused me and other devs

**Comment by 4ian** [comment]
Please don't fight on an issue though ;)
I think there is good intent from everyone, this interface is maybe a bit confusing even if working as intended (or maybe there is a better way to make it work).
Maybe some clearer explanations somewhere?  

> How do you expect us to show each path of 25 frames? Inside piskel.

I think we're hitting the limits of the fact that Piskel is just "pasted" on top of GDevelop. It may not be the better integration on this point. This being said, not sure if it's worth spending more time making this integration more complex. 
I would rather just make things more obvious for the user about what they are editing and call it a day.

**Comment by blurymind** [comment]
Yes indeed, it has limits and introducing more logic to this could make it even more confusing.
I dont think we have a proper documentation for piskel inside gdevelop at the wiki. Would that be a satisfactory way to address this?
